// frontend/src/store/useStore.ts
/**
 * CAREHIVE global state (Zustand)
 */
import { create } from 'zustand';
import type { UserRole, RolePermissions } from '@/types/roles';
import { ROLE_PERMISSIONS } from '@/types/roles';
import type { Medication, MedicationLog } from '@/types/medication';

export interface Intervention {
  id: string;
  userId: string;
  agentType: string;
  message: string;
  priority: string;
  createdAt: string;
  read?: boolean;
}

export interface PatientSummary {
  id: string;
  name: string;
  age: number;
  condition: string;
}

interface CarehiveState {
  userId: string | null;
  userName: string | null;
  userRole: UserRole;
  setUser: (id: string, name: string | null) => void;
  setUserRole: (role: UserRole) => void;
  permissions: RolePermissions;

  /** For clinician/family: the patient whose data we're currently viewing */
  viewingPatientId: string | null;
  viewingPatientName: string | null;
  setViewingPatient: (id: string | null, name: string | null) => void;
  patientList: PatientSummary[];
  setPatientList: (list: PatientSummary[]) => void;

  accessToken: string | null;
  setAccessToken: (token: string | null) => void;
  authReady: boolean;
  setAuthReady: (ready: boolean) => void;

  interventions: Intervention[];
  addIntervention: (i: Intervention) => void;
  setInterventions: (i: Intervention[]) => void;

  medications: Medication[];
  setMedications: (meds: Medication[]) => void;
  addMedication: (med: Medication) => void;
  updateMedication: (id: string, updates: Partial<Medication>) => void;
  removeMedication: (id: string) => void;

  medicationLogs: MedicationLog[];
  setMedicationLogs: (logs: MedicationLog[]) => void;
  addMedicationLog: (log: MedicationLog) => void;
}

export const useStore = create<CarehiveState>((set) => ({
  userId: null,
  userName: null,
  userRole: 'patient' as UserRole,
  permissions: ROLE_PERMISSIONS.patient,
  setUser: (id, name) => set({ userId: id, userName: name }),
  setUserRole: (role) => set({ userRole: role, permissions: ROLE_PERMISSIONS[role] }),

  viewingPatientId: null,
  viewingPatientName: null,
  setViewingPatient: (id, name) => set({ viewingPatientId: id, viewingPatientName: name }),
  patientList: [],
  setPatientList: (patientList) => set({ patientList }),

  accessToken: null,
  setAccessToken: (token) => set({ accessToken: token }),
  authReady: false,
  setAuthReady: (ready) => set({ authReady: ready }),

  interventions: [],
  addIntervention: (i) =>
    set((s) => ({
      interventions: [
        { ...i, createdAt: i.createdAt || new Date().toISOString() },
        ...s.interventions,
      ],
    })),
  setInterventions: (interventions) => set({ interventions }),

  medications: [],
  setMedications: (medications) => set({ medications }),
  addMedication: (med) => set((s) => ({ medications: [med, ...s.medications] })),
  updateMedication: (id, updates) =>
    set((s) => ({
      medications: s.medications.map((m) => (m.id === id ? { ...m, ...updates } : m)),
    })),
  removeMedication: (id) =>
    set((s) => ({ medications: s.medications.filter((m) => m.id !== id) })),

  medicationLogs: [],
  setMedicationLogs: (medicationLogs) => set({ medicationLogs }),
  addMedicationLog: (log) => set((s) => ({ medicationLogs: [log, ...s.medicationLogs] })),
}));
