// frontend/src/components/AuthLoader.tsx
'use client';

import { useEffect, useRef } from 'react';
import { supabase } from '@/lib/supabase';
import { api } from '@/lib/api';
import { useStore } from '@/store/useStore';
import { subscribeToUser } from '@/lib/socket';
import type { UserRole } from '@/types/roles';

export function AuthLoader() {
  const { setUser, setAccessToken, setAuthReady, setUserRole } = useStore();
  const done = useRef(false);
  const roleApplied = useRef(false);

  async function syncUser(session: { access_token: string; user?: { id?: string; email?: string | null } }) {
    setAccessToken(session.access_token);
    try {
      const u = await api.getMe();
      setUser(u.id, u.name);

      if (roleApplied.current) return;

      const pendingRole = typeof window !== 'undefined'
        ? localStorage.getItem('carehive_signup_role') as UserRole | null
        : null;

      const needsRoleUpdate = pendingRole
        && ['patient', 'clinician', 'family'].includes(pendingRole)
        && pendingRole !== 'patient'
        && (!u.role || u.role === 'patient');

      if (needsRoleUpdate) {
        roleApplied.current = true;
        setUserRole(pendingRole);
        localStorage.removeItem('carehive_signup_role');

        try {
          const updated = await api.updateProfile({
            name: u.name || 'User',
            age: u.age || 1,
            condition: u.condition || 'other',
            role: pendingRole,
          });
          if (updated.role && updated.role !== 'patient') {
            setUserRole(updated.role as UserRole);
          }
        } catch (e) {
          console.warn('Failed to persist role to backend:', e);
        }
      } else {
        const dbRole = (u.role || 'patient') as UserRole;
        if (dbRole !== 'patient' || !pendingRole) {
          setUserRole(dbRole);
        }
        if (pendingRole) localStorage.removeItem('carehive_signup_role');
      }

      subscribeToUser(u.id);
    } catch {
      setUser(session.user?.id ?? '', session.user?.email ?? null);
      subscribeToUser(session.user?.id ?? '');
    }
  }

  useEffect(() => {
    if (done.current) return;
    done.current = true;

    (async () => {
      if (supabase) {
        const {
          data: { session },
        } = await supabase.auth.getSession();
        if (session?.access_token) {
          await syncUser(session);
          setAuthReady(true);
          return;
        }
        setAccessToken(null);
        setUser('', null);
      }
      setAuthReady(true);
    })();
  }, [setUser, setAccessToken, setAuthReady, setUserRole]);

  useEffect(() => {
    if (!supabase) return;
    const {
      data: { subscription },
    } = supabase.auth.onAuthStateChange(async (_, session) => {
      if (session?.access_token) {
        await syncUser(session);
        setAuthReady(true);
      } else {
        setAccessToken(null);
        setUser('', null);
        roleApplied.current = false;
      }
    });
    return () => subscription.unsubscribe();
  }, [setUser, setAccessToken, setUserRole]);

  return null;
}
