// frontend/src/app/profile/page.tsx
'use client';

import { useEffect, useState } from 'react';
import { useRouter } from 'next/navigation';
import Link from 'next/link';
import { supabase } from '@/lib/supabase';
import { api } from '@/lib/api';
import { useStore } from '@/store/useStore';
import { signOut } from '@/lib/auth';
import { Card, CardContent, CardHeader, CardTitle } from '@/components/ui/Card';
import { Button } from '@/components/ui/Button';
import { Badge } from '@/components/ui/Badge';
import { motion } from 'framer-motion';
import type { UserRole } from '@/types/roles';
import { ROLE_LABELS, ROLE_DESCRIPTIONS } from '@/types/roles';

const CONDITIONS = ['diabetes', 'hypertension', 'copd', 'heart_failure', 'other'] as const;

const ROLE_BADGE_VARIANT: Record<UserRole, 'success' | 'info' | 'warning'> = {
  patient: 'success',
  clinician: 'info',
  family: 'warning',
};

export default function ProfilePage() {
  const router = useRouter();
  const { accessToken, setUser, setUserRole, userRole } = useStore();
  const [name, setName] = useState('');
  const [age, setAge] = useState('');
  const [condition, setCondition] = useState('');
  const [role, setRole] = useState<UserRole>(userRole);
  const [loading, setLoading] = useState(false);
  const [error, setError] = useState<string | null>(null);
  const [saved, setSaved] = useState(false);

  const isNew = typeof window !== 'undefined' && new URLSearchParams(window.location.search).get('new') === '1';
  const signupRole = typeof window !== 'undefined' ? localStorage.getItem('carehive_signup_role') as UserRole | null : null;

  useEffect(() => {
    if (!accessToken) return;
    api
      .getMe()
      .then((u) => {
        setName(u.name);
        setAge(String(u.age));
        setCondition(u.condition);
        const storeRole = useStore.getState().userRole;
        if (storeRole && storeRole !== 'patient') {
          setRole(storeRole);
        } else if (u.role) {
          setRole(u.role as UserRole);
        }
      })
      .catch(() => {});
  }, [accessToken]);

  async function handleSubmit(e: React.FormEvent) {
    e.preventDefault();
    if (!accessToken) {
      router.push('/login');
      return;
    }
    const ageNum = parseInt(age, 10);
    if (!name.trim() || isNaN(ageNum) || ageNum < 1 || ageNum > 120 || !condition) {
      setError('Please enter name, valid age (1-120), and condition.');
      return;
    }
    setError(null);
    setLoading(true);
    try {
      const storeRole = useStore.getState().userRole;
      const roleToSave = isNew && signupRole ? signupRole : storeRole;
      const u = await api.updateProfile({ name: name.trim(), age: ageNum, condition, role: roleToSave });
      setUser(u.id, u.name);
      if (u.role) setUserRole(u.role as UserRole);
      if (isNew) localStorage.removeItem('carehive_signup_role');
      setSaved(true);
      if (isNew) setTimeout(() => router.push('/dashboard'), 1500);
    } catch (err) {
      setError(err instanceof Error ? err.message : 'Failed to save');
    } finally {
      setLoading(false);
    }
  }

  function handleSignOut() {
    signOut(router);
  }

  async function handleDeleteAccount() {
    if (!confirm('Are you sure you want to delete your account? This cannot be undone.')) return;
    try {
      await api.deleteAccount();
    } catch {
      // proceed anyway
    }
    localStorage.removeItem('carehive_signup_role');
    signOut(router);
  }

  if (!supabase) {
    return (
      <div className="p-8 max-w-lg mx-auto">
        <p className="text-slate-600 dark:text-slate-300">Auth is not configured.</p>
        <Link href="/" className="mt-4 inline-block text-teal-600 hover:underline dark:text-teal-400">
          Back to home
        </Link>
      </div>
    );
  }

  if (!accessToken) {
    return (
      <div className="p-8 max-w-lg mx-auto">
        <Card>
          <CardContent className="pt-6">
            <p className="text-slate-600 dark:text-slate-300">Sign in to manage your profile.</p>
            <Link href="/login" className="mt-4 inline-block text-teal-600 hover:underline">
              Go to Sign in
            </Link>
          </CardContent>
        </Card>
      </div>
    );
  }

  const displayRole = isNew && signupRole ? signupRole : (userRole !== 'patient' ? userRole : role);
  const inputClass =
    'w-full rounded-lg border border-slate-200 dark:border-slate-700 bg-white dark:bg-slate-900 px-3 py-2.5 text-sm text-slate-900 dark:text-slate-100 focus:ring-2 focus:ring-teal-500/50 focus:border-teal-500 outline-none transition-colors';

  return (
    <motion.div
      initial={{ opacity: 0, y: 10 }}
      animate={{ opacity: 1, y: 0 }}
      transition={{ duration: 0.35 }}
      className="p-6 md:p-8 max-w-lg mx-auto"
    >
      <Card>
        <CardHeader className="flex flex-row items-center justify-between">
          <CardTitle>Your Profile</CardTitle>
          <button
            type="button"
            onClick={handleSignOut}
            className="text-sm text-slate-500 hover:text-slate-700 dark:text-slate-400 dark:hover:text-slate-200"
          >
            Sign out
          </button>
        </CardHeader>
        <CardContent>
          {isNew && (
            <div className="text-sm text-teal-700 dark:text-teal-300 bg-teal-50 dark:bg-teal-950/30 border border-teal-100 dark:border-teal-800 rounded-lg px-3 py-2 mb-4">
              Complete your profile so CAREHIVE can personalize your care plan.
            </div>
          )}
          <form onSubmit={handleSubmit} className="space-y-4">
            {error && (
              <p className="text-sm text-red-600 dark:text-red-400 bg-red-50 dark:bg-red-950/30 border border-red-100 dark:border-red-800 rounded-lg px-3 py-2">
                {error}
              </p>
            )}
            {saved && (
              <p className="text-sm text-green-700 dark:text-green-300 bg-green-50 dark:bg-green-950/30 border border-green-100 dark:border-green-800 rounded-lg px-3 py-2">
                Profile saved.
              </p>
            )}
            <div>
              <label className="block text-sm font-medium text-slate-700 dark:text-slate-300 mb-1">Name</label>
              <input type="text" value={name} onChange={(e) => setName(e.target.value)} required className={inputClass} />
            </div>
            <div>
              <label className="block text-sm font-medium text-slate-700 dark:text-slate-300 mb-1">Age</label>
              <input
                type="number"
                min={1}
                max={120}
                value={age}
                onChange={(e) => setAge(e.target.value)}
                required
                className={inputClass}
              />
            </div>
            <div>
              <label className="block text-sm font-medium text-slate-700 dark:text-slate-300 mb-1">Condition</label>
              <select value={condition} onChange={(e) => setCondition(e.target.value)} required className={inputClass}>
                <option value="">Select...</option>
                {CONDITIONS.map((c) => (
                  <option key={c} value={c}>
                    {c.replace('_', ' ')}
                  </option>
                ))}
              </select>
            </div>

            {/* Role — read-only display */}
            <div>
              <label className="block text-sm font-medium text-slate-700 dark:text-slate-300 mb-2">Role</label>
              <div className="flex items-center gap-3 p-3 rounded-lg border border-slate-200 dark:border-slate-700 bg-slate-50 dark:bg-slate-800/50">
                <Badge variant={ROLE_BADGE_VARIANT[displayRole]} size="sm">{ROLE_LABELS[displayRole]}</Badge>
                <span className="text-xs text-slate-500 dark:text-slate-400">{ROLE_DESCRIPTIONS[displayRole]}</span>
              </div>
              <p className="text-xs text-slate-400 mt-1">Role is set during registration and cannot be changed.</p>
            </div>

            <Button type="submit" disabled={loading} className="w-full">
              {loading ? 'Saving...' : 'Save Profile'}
            </Button>
          </form>
          <div className="flex items-center justify-between mt-4">
            <Link
              href="/dashboard"
              className="text-sm text-teal-600 hover:underline dark:text-teal-400"
            >
              Back to Dashboard
            </Link>
            <button
              type="button"
              onClick={handleDeleteAccount}
              className="text-sm text-red-500 hover:text-red-700 dark:text-red-400 dark:hover:text-red-300"
            >
              Delete Account
            </button>
          </div>
        </CardContent>
      </Card>
    </motion.div>
  );
}
