// frontend/src/lib/api.ts
/**
 * CAREHIVE API client.
 * When accessToken is set in store, requests include Authorization: Bearer for real user data.
 */
import { useStore } from '@/store/useStore';
import type { Medication, MedicationLog, DailyScheduleItem } from '@/types/medication';
import type { EhrRecord } from '@/types/ehr';
import type { UserRole } from '@/types/roles';

const API_BASE = process.env.NEXT_PUBLIC_API_URL || 'http://localhost:4000/api';

async function fetchApi<T>(path: string, options?: RequestInit): Promise<T> {
  const token = useStore.getState().accessToken;
  const headers: Record<string, string> = {
    'Content-Type': 'application/json',
    ...(options?.headers as Record<string, string>),
  };
  if (token) headers['Authorization'] = `Bearer ${token}`;
  const res = await fetch(`${API_BASE}${path}`, { ...options, headers });
  if (!res.ok) throw new Error(await res.text().catch(() => res.statusText));
  return res.json() as Promise<T>;
}

export interface UserMe {
  id: string;
  name: string;
  age: number;
  condition: string;
  email?: string;
  role?: UserRole;
  createdAt?: string;
}

export const api = {
  // --- User ---
  getDemoUser: () => fetchApi<UserMe>('/user/demo'),
  getMe: () => fetchApi<UserMe>('/user/me'),
  updateProfile: (body: { name: string; age: number; condition: string; role?: string }) =>
    fetchApi<UserMe>('/user/profile', { method: 'POST', body: JSON.stringify(body) }),
  getPatients: () =>
    fetchApi<{ patients: Array<{ id: string; name: string; age: number; condition: string }> }>('/user/patients'),
  deleteAccount: () =>
    fetchApi<{ success: boolean }>('/user/me', { method: 'DELETE' }),

  // --- Health ---
  logHealth: (body: { userId?: string; medicationTaken: boolean; steps: number; mood: number; notes?: string }) =>
    fetchApi<{ id: string }>('/health/log', { method: 'POST', body: JSON.stringify(body) }),
  getHistory: (userId: string, limit = 30) =>
    fetchApi<Array<{ id: string; medicationTaken: boolean; steps: number; mood: number; timestamp: string }>>(
      `/health/history?userId=${encodeURIComponent(userId)}&limit=${limit}`
    ),

  // --- Insights ---
  getInsights: (userId: string) =>
    fetchApi<{
      insights: Array<{ type: string; title: string; description: string; severity: string; agentSource?: string }>;
    }>(`/insights?userId=${encodeURIComponent(userId)}`),
  runOrchestrator: (userId: string) =>
    fetchApi<{ summary: string; interventionsCreated: number }>(
      `/insights/orchestrate?userId=${encodeURIComponent(userId)}`,
      { method: 'POST' }
    ),

  // --- Clinician ---
  getClinicianSummary: (userId: string) =>
    fetchApi<{
      userName: string;
      condition: string;
      periodStart: string;
      periodEnd: string;
      adherenceRate: number;
      riskFlags: string[];
      trends: string[];
      agentHighlights: string[];
      generatedAt: string;
    }>(`/clinician/summary?userId=${encodeURIComponent(userId)}`),

  // --- Agents ---
  getAgentActivity: (userId: string) =>
    fetchApi<{ activity: Array<{ agentType: string; message: string; priority: string; timestamp: string }> }>(
      `/agents/activity?userId=${encodeURIComponent(userId)}`
    ),

  // --- Chat ---
  chat: (message: string, userId?: string) =>
    fetchApi<{ reply: string }>('/chat', { method: 'POST', body: JSON.stringify({ message, userId }) }),

  // --- Vision ---
  visionAnalyze: (body: { image: string; mimeType?: string }) =>
    fetchApi<{ analysis: { mood: number; posture: string; summary?: string }; healthLog?: unknown }>(
      '/vision/analyze',
      { method: 'POST', body: JSON.stringify(body) }
    ),

  // --- Medications ---
  getMedications: (activeOnly = true) =>
    fetchApi<{ medications: Medication[] }>(`/medications?active=${activeOnly}`),
  addMedication: (body: {
    name: string;
    dosage: string;
    frequency: string;
    timing: string[];
    instructions?: string;
  }) => fetchApi<Medication>('/medications', { method: 'POST', body: JSON.stringify(body) }),
  updateMedication: (id: string, body: Partial<Medication>) =>
    fetchApi<Medication>(`/medications/${id}`, { method: 'PUT', body: JSON.stringify(body) }),
  deleteMedication: (id: string) =>
    fetchApi<{ success: boolean }>(`/medications/${id}`, { method: 'DELETE' }),
  logMedicationDose: (body: { medicationId: string; status: string; scheduledTime: string }) =>
    fetchApi<MedicationLog>('/medications/log', { method: 'POST', body: JSON.stringify(body) }),
  getMedicationLogs: (date?: string) =>
    fetchApi<{ logs: MedicationLog[] }>(`/medications/logs${date ? `?date=${date}` : ''}`),
  getDailySchedule: (date?: string) =>
    fetchApi<{ schedule: DailyScheduleItem[]; date: string }>(
      `/medications/schedule${date ? `?date=${date}` : ''}`
    ),

  // --- EHR ---
  uploadEhr: (body: { fileName: string; rawText: string; patientId?: string }) =>
    fetchApi<EhrRecord>('/ehr/upload', { method: 'POST', body: JSON.stringify(body) }),
  getEhrRecords: (patientId?: string) =>
    fetchApi<{ records: EhrRecord[] }>(`/ehr${patientId ? `?patientId=${patientId}` : ''}`),
  getEhrRecord: (id: string) => fetchApi<EhrRecord>(`/ehr/${id}`),
  convertEhrToMedications: (ehrRecordId: string) =>
    fetchApi<{ created: number; medications: Medication[] }>('/ehr/convert-medications', {
      method: 'POST',
      body: JSON.stringify({ ehrRecordId }),
    }),
};
